stan/math/prim/fun/vec_concat.hpp

```diff
@@ -7,31 +7,55 @@
 
 namespace stan {
 namespace math {
-
+namespace internal {
+inline auto sum_vector_sizes() { return 0; }
+/**
+ * Get the internal sizes of a pack of vectors.
+ * @tparam Vec Type of first vector to count.
+ * @tparam VecArgs Parameter pack of vector types to accumulate sizes of
+ * @param x vector to start accumulation of sizes with.
+ * @param args pack of vectors to accumulate sizes for.
+ */
+template <typename Vec, typename... VecArgs>
+inline auto sum_vector_sizes(Vec&& x, VecArgs&&... args) {
+  return x.size() + sum_vector_sizes(std::forward<VecArgs>(args)...);
+}
+/**
+ * End of recursion for appending to vector.
+ */
+template <typename VecInOut>
+inline void append_vectors(VecInOut& x) {}
 /**
- * Ends the recursion to extract the event stack.
- * @param v1 Events on the OpenCL event stack.
- * @return returns input vector.
+ * Fill a vector with other vectors.
+ * @tparam VecInOut Type of vector to be filled.
+ * @tparam VecIn Type of the first vector to insert into the in/out vector.
+ * @tparam VecArgs Parameter pack of other vectors to fill in the in/out vector.
+ * @param x Vector to be filled.
+ * @param y First vector to insert into the in/out vector.
+ * @param args Pack of other vectors to fill in the in/out vector.
  */
-template <typename T>
-inline const std::vector<T>& vec_concat(const std::vector<T>& v1) {
-  return v1;
+template <typename VecInOut, typename VecIn, typename... VecArgs>
+inline void append_vectors(VecInOut& x, VecIn&& y, VecArgs&&... args) {
+  x.insert(x.end(), y.begin(), y.end());
+  append_vectors(x, std::forward<VecArgs>(args)...);
 }
+}  // namespace internal
 
 /**
  * Get the event stack from a vector of events and other arguments.
  *
- * @tparam T type of elements in the array
+ * @tparam Vec type of elements in the array
  * @tparam Args types for variadic arguments
  * @param v1 event stack to roll up
  * @param args variadic arguments passed down to the next recursion
  * @return Vector of OpenCL events
  */
-template <typename T, typename... Args>
-inline const std::vector<T> vec_concat(const std::vector<T>& v1,
-                                       const Args... args) {
-  std::vector<T> vec = vec_concat(args...);
-  vec.insert(vec.end(), v1.begin(), v1.end());
+template <typename Vec, typename... Args>
+inline auto vec_concat(Vec&& v1, Args&&... args) {
+  std::vector<value_type_t<Vec>> vec;
+  vec.reserve(internal::sum_vector_sizes(v1, args...));
+  internal::append_vectors(vec, std::forward<Vec>(v1),
+                           std::forward<Args>(args)...);
```

Comment:
Why do you need to change this implementation? As I see it does the same. If there is a good argument, remove the forwarding here too.

Reply:
> Why do you need to change this implementation? 

It doesn't need to be changed, but this one will be just a hair better since we allocate all the space up-front instead of using `push_back()` and resizing.

> As I see it does the same. If there is a good argument, remove the forwarding here too.

Yeah I'll change this, I didn't read the docs for `vector.insert()` which doesn't do anything with forwarding references